src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+                if (this.storeTip != this.chain.Genesis)
+                {
+                    this.logger.LogTrace("(-)[REBUILD_REQUIRED]");
+                    throw new BlockStoreException("You need to rebuild the block store database using -reindex-chainstate to change -txindex");
+                }
+
+                if (this.storeSettings.TxIndex)
+                    await this.blockRepository.SetTxIndexAsync(this.storeSettings.TxIndex).ConfigureAwait(false);
+            }
+            
+            // Throw if block store was initialized after the consensus.
+            // This is needed in order to rewind consensus in case it is initialized ahead of the block store.
+            if (this.chainState.ConsensusTip != null)
+            {
+                this.logger.LogCritical("Block store initialized after the consensus!");
+                this.logger.LogTrace("(-)[INITIALIZATION_ERROR]");
+                throw new BlockStoreException("Block store initialized after consensus!");
+            }
+
+            // Start dequeuing.
+            this.currentBatchSizeBytes = 0;
+            this.dequeueLoopTask = this.DequeueBlocksContinuouslyAsync();
+
+            this.logger.LogTrace("(-)");
+        }
+        
+        /// <summary>Sets the internal store tip and exposes the store tip to other components through the chain state.</summary>
+        private void SetStoreTip(ChainedHeader newTip)
+        {
+            this.storeTip = newTip;
+            this.chainState.BlockStoreTip = newTip;
+        }
+
+        /// <summary>
+        /// Sets block store tip to the last block that exists both in the repository and in the <see cref="ConcurrentChain"/>.
+        /// </summary>
+        private async Task RecoverStoreTipAsync()
+        {
+            this.logger.LogTrace("()");
+
+            var blockStoreResetList = new List<uint256>();
+            uint256 resetBlockHash = this.blockRepository.BlockHash;
+            Block resetBlock = await this.blockRepository.GetAsync(resetBlockHash).ConfigureAwait(false);
+
+            while (this.chain.GetBlock(resetBlockHash) == null)
+            {
+                blockStoreResetList.Add(resetBlockHash);
+
+                if (resetBlock.Header.HashPrevBlock == this.chain.Genesis.HashBlock)
+                {
+                    resetBlockHash = this.chain.Genesis.HashBlock;
+                    break;
+                }
+
+                resetBlock = await this.blockRepository.GetAsync(resetBlock.Header.HashPrevBlock).ConfigureAwait(false);
+
+                if (resetBlock == null)
+                {
+                    // This can happen only if block store is corrupted.
+                    throw new BlockStoreException("Block store failed to recover.");
+                }
+
+                resetBlockHash = resetBlock.GetHash();
+            }
+
+            ChainedHeader newTip = this.chain.GetBlock(resetBlockHash);
+
+            if (blockStoreResetList.Count != 0)
```

Comment:
Just wondering why not use `Any()`

Reply:
Fixed since it looks nicer that way. 

But I Any() have more overhead
https://stackoverflow.com/questions/18393766/how-does-linq-any-works-internally

Reply:
PLS REVERT BACK

Reply:
Reverted